Onboarding: Add profiler body class on initial load + respect skipped/completed flags

justinshreve opened the pull request:
This PR adds the `woocommerce-profile-wizard__body` body class on initial load of the PHP app if we know that the profiler will display and if the user is on the main JS page.

This allows us to prevent the initial flash of the sidebar / masterbar (which makes for a smoother transition between Calypso and `wc-admin`).

It also adds a (temporary) PHP flag so that we can enable/disable the profiler, without editing `dashboard/index.js`. (I don't know about you Josh, but it was getting annoying to switch branches, stash the change, etc 😆). In addition to that, I've added some logic so the profiler will hide/show based on the skipped/complete flags.

To Test:
* Set `WOOCOMMERCE_ADMIN_DEV_SHOW_PROFILER` to `true` wherever you keep your development snippets / constants.
* Make sure `skipped` and `completed` are both false via `/wc-admin/v1/onboarding/profile`
* Visit `/wp-admin/admin.php?page=wc-admin#/`. Notice that the admin menus do not display, and the profiler loads in (Another good example of this is running through the plugins flow and seeing the transition between Calypso).
* Test setting `skipped` or `completed` to true, and refresh the page. Note that the profiler does not show.
* Test visiting an analytics route, and make sure the admin bar loads in correctly.

timmyc commented on includes/features/onboarding/class-wc-admin-onboarding.php:
```diff
@@ -26,6 +26,65 @@ public static function get_instance() {
+	 * @param string $admin_body_class Body class to add.
+	 */
+	public function add_admin_body_classes( $admin_body_class = '' ) {
+		global $hook_suffix;
```
why is this global needed?

justinshreve replied:
Oops, good catch. It was a copy and paste from the body class function in `WC_Admin_Loader`. I'll remove it.

joshuatf commented on includes/features/onboarding/class-wc-admin-onboarding.php:
```diff
@@ -26,6 +26,65 @@ public static function get_instance() {
+	 */
+	public function should_show_profiler() {
+		// @todo Remove this once we have a proper way to dismiss the profiler.
+		if ( ! defined( 'WOOCOMMERCE_ADMIN_DEV_SHOW_PROFILER' ) || false === WOOCOMMERCE_ADMIN_DEV_SHOW_PROFILER ) {
```
Yesss!  💯

joshuatf commented on includes/features/onboarding/class-wc-admin-onboarding.php:
```diff
@@ -26,6 +26,65 @@ public static function get_instance() {
+		// @todo Update this to compare to proper bools (https://github.com/woocommerce/woocommerce-admin/issues/2299).
+		$is_completed = isset( $onboarding_data['completed'] ) && 'true' === $onboarding_data['completed'];
+		$is_skipped   = isset( $onboarding_data['skipped'] ) && 'true' === $onboarding_data['skipped'];
+		return $is_completed || $is_skipped ? false : true;
```
What happens if a user has already gone through the default WC setup wizard?  Do they get hit with this again or do we want to add a note to change this before merging into core?

justinshreve replied:
Good question. We should probably respect the current flag.

justinshreve replied:
There actually isn't a flag for this exactly. The redirect and notice both just rely on the presence of the `install` notice showing. 

https://github.com/woocommerce/woocommerce/blob/af574f6484a4300d1ec92c569593013f2e2fced2/includes/admin/class-wc-admin.php#L161-L164

https://github.com/woocommerce/woocommerce/blob/d0491072e8a7d0a4535db0ec2bae4a6d43ff64fd/includes/admin/views/html-notice-install.php

I think instead of relying on the old notice, we should probably add a note and/or an issue here on GitHub with the label 'On Core Merge'.

I'm thinking when we merge onboarding to core, we could set the completed flag to true during the upgrade process.

joshuatf replied:
Sounds like a great solution to me!

joshuatf commented on includes/api/class-wc-admin-rest-onboarding-plugins-controller.php:
```diff
@@ -206,13 +206,14 @@ public function connect_jetpack() {
 		}
 
-		$redirect_url = esc_url_raw(
+		$next_step_slug = 'store-details';
```
Side note: I know we talked a bit about extensibility before.  Would we ever want to allow hooking into change this step?

joshuatf commented on client/layout/index.js:
```diff
@@ -27,6 +27,12 @@ import StoreAlerts from './store-alerts';
+		const path = this.getPath();
+
+		if ( 'dashboard' !== path ) {
+			document.body.classList.remove( 'woocommerce-profile-wizard__body' );
```
Would it be possible to add a check in `add_admin_body_classes()` to see if we're in the dashboard page before adding this class?  That way we can remove these 3 lines and keep this components scope down.

justinshreve replied:
I have a check in there that checks if we are on a JS page, but the problem is we can't tell which wc-admin route they on because the route is stored in a JS hash that is not passed down to the server.

I did a little exploration to see what we can do, but switching to another routing method / URL structure would be non-trivial. I think we might have to stick with this for now. Does that seem OK?

justinshreve replied:
https://github.com/woocommerce/woocommerce-admin/issues/2267

joshuatf replied:
Ya, that seems very reasonable and thanks for your research around the issue!  If we make those routing changes you suggested at a later date, we can revisit this.

justinshreve replied:
I ended up generalizing this a bit more in the meantime. I now have it adding a `woocommerce-admin-is-loading` class.

I'm still only appending it when necessary, but this allows us to do something similar on other routes in the future (like future Woo App pages). It also allows me to drop the `dashboard` check here. We'll just always remove the class once `<Layout />` has loaded, and the  profiler will append its own profiler class to continue hiding elements.

justinshreve commented:
Thanks for the review! I've addressed some of the feedback and added some todos/notes for the future. https://github.com/woocommerce/woocommerce-admin/pull/2300/commits/9f636401902a144b723b5074508feb0ab5ddb9d5
